File: 剑指offer/46.把数字翻译成字符串.py
class Solution:
    def translateNum(self, num: int) -> int:
        s = str(num)
        a = 1
        b = 1

        for i in range(2, len(s) + 1):
            tmp = s[i - 2:i]
            c = a + b if '10' <= tmp <= '25' else b
            a = b
            b = c

        return b
        

File: 剑指offer/46.把数字翻译成字符串.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int translateNum(int num)
    {
        string nums = to_string(num);
        int n = nums.length();

        vector<int> dp(n + 1, 0);
        int p = 1;
        int q = 1;

        for (int i = 2; i <= n; i++) {
            int r = q;
            if (nums.substr(i - 2, 2) >= "10" && nums.substr(i - 2, 2) <= "25") {
                r += p;
            }
            p = q;
            q = r;
        }

        return q;
    }
};